File: List_4_C/List_4.scala
sealed trait BT[A]
  case class Empty[A]() extends BT[A]
  case class Node[A](elem:A, left:BT[A], right:BT[A]) extends BT[A]

sealed trait Graphs[A]
  case class Graph[A](succ: A=>List[A]) extends Graphs[A]


object List_4 {
  val tt = Node[Int](1,
    Node[Int](2,
      Node[Int](4,
        Empty[Int],
        Empty[Int]
      ),
      Empty[Int]
    ),
    Node[Int](3,
      Node[Int](5,
        Empty[Int],
        Node[Int](6,
          Empty[Int],
          Empty[Int]
        )
      ),
      Empty[Int]
    )
  )
  val g = Graph((i: Int) => i match {
    case 0 => List(3)
    case 1 => List(0,2,4)
    case 2 => List(1)
    case 3 => Nil
    case 4 => List(0,2)
    case n => throw new Exception("Graph g: node „ + n" + " doesn't exist")
  })

  // Zad 3
  def breadthBT[T](bt: BT[T]): List[T] = {
    def breadthIn(acc: List[BT[T]], res: List[T]): List[T] = {
      if (acc == Nil) {
        res
      } else {
        acc.head match {
          case Empty() => breadthIn(acc.tail, res)
          case Node(x, l, r) => breadthIn(acc.tail ++ List(l, r), x :: res)
        }
      }
    }
    breadthIn(List(bt), List.empty[T]).reverse
  }

  // Zad 4
  def internalPath(bt: BT[Int]): Int = {
    def internal(n: BT[Int], res: Int, lvl: Int): Int =
      n match {
        case Empty() => res
        case Node(v, l, r) => internal(l, internal(r, res + lvl, lvl + 1), lvl + 1)
      }
    internal(bt, 0, 0)
  }

  def externalPath(bt: BT[Int]) = {
    def extern(n: BT[Int], res: Int, lvl: Int): Int = {
      n match {
        case Empty() => res + lvl
        case Node(v, l, r) => extern(l, extern (r,res, lvl+1), lvl+1)
      }
    }
    extern(bt, 0, 0)
  }

  // Zad 5
  def depthSearch(gr: Graph[Int], st: Int):List[Int] = {
    def dfs(acc:List[Int], res:List[Int]):List[Int] = {
      if (acc == Nil) {
        res
      } else {
        if (res.contains(acc.head)) {
          dfs(acc.tail, res)
        } else {
          dfs((gr succ acc.head)++acc.tail, acc.head::res)
        }
      }
    }
    dfs(List(st), List.empty[Int]).reverse
  }
}

File: Semestr_3/Paradigms/List_4_C/List_4.ml
type 'a bt = Empty | Node of 'a * 'a bt * 'a bt;;
  
let tt = Node(1, Node(2, Node(4, Empty, Empty), Empty ), Node(3,
  Node(5, Empty, Node(6, Empty, Empty)), Empty));;
 
 let breadthBT bt =
    let rec breadthIN acc res =
        if acc = [] then  res
        else 
            match List.hd acc with
            Empty -> breadthIN (List.tl acc) res
            | Node(v, l, r) -> breadthIN ((List.tl acc) @ [l; r]) (v::res) in
    List.rev (breadthIN [bt] []);;
    
let internalPath bt =
    let rec internal n res lvl =
    match n with
        Empty -> res
        | Node(v, l, r) -> internal l (internal r (res+lvl) (lvl+1)) (lvl+1) in
    internal bt 0 0;;

let externalPath bt =
    let rec extern n res lvl =
    match n with
        Empty -> res + lvl
        | Node(v, l, r) -> extern l (extern r res (lvl+1)) (lvl+1) in
    extern bt 0 0;;

type 'a graph = Graph of ('a -> 'a list);;
 let g = Graph
 (function
 0 -> [3]
 | 1 -> [0;2;4]
 | 2 -> [1]
 | 3 -> []
 | 4 -> [0;2]
 | n -> failwith ("Graph g: node "^string_of_int n^" doesn't exist")
);;

let depthSearch (Graph gr) st =
    let rec dfs acc res =
        if acc = [] then res
        else 
            if  List.mem (List.hd acc) res then
                dfs (List.tl acc) res
            else 
                dfs ((gr (List.hd acc)) @ (List.tl acc)) ((List.hd acc) :: res)  in
    List.rev (dfs [st] []);;
